File: NoTomorrow/Features/Bro/CantMakeItSheet.swift
import SwiftUI
import SwiftData

/// "Can't make it today" sheet, opened from the Dashboard and the Bro tab.
/// Sending writes my `AttendanceRecord(.cancelled)`, a `HeadsUp(.cantMakeIt)` and, when paired, tells the partner.
struct CantMakeItSheet: View {
    var sessionDay: Date
    var minuteOfDay: Int
    var routineName: String?
    var onDone: () -> Void

    @Environment(\.dismiss) private var dismiss
    @Environment(\.modelContext) private var modelContext
    @Query private var pairings: [BroPairing]
    @Query private var schedules: [GymSchedule]

    @State private var reason: CantReason?
    @State private var note = ""
    @State private var makeUp: MakeUpChoice?
    @State private var isSending = false
    @FocusState private var noteFocused: Bool

    private let maxNoteLength = 80

    enum CantReason: String, CaseIterable, Identifiable {
        case sick, work, tired, family, none
        var id: String { rawValue }
        var key: LocalizedStringKey { LocalizedStringKey(String("cant.reason.\(rawValue)")) }
        var label: String { String(localized: String.LocalizationValue("cant.reason.\(rawValue)")) }
    }

    enum MakeUpChoice: Equatable {
        case day(Date)
        case skip
        var date: Date? { if case .day(let d) = self { return d } else { return nil } }
    }

    private var partnerName: String? {
        BroShared.service.partner?.name ?? pairings.first?.partnerName
    }
    private var isPaired: Bool { partnerName != nil }
    private var makeUpDays: [Date] { BroDerived.nonGymDays(after: sessionDay, schedule: schedules.first) }

    var body: some View {
        VStack(spacing: 0) {
            Grabber()
            ScrollView {
                VStack(alignment: .leading, spacing: 18) {
                    titleBlock
                    whySection
                    makeUpSection
                    summaryBox
                    buttons
                }
                .padding(.horizontal, NT.Spacing.screenH)
                .padding(.top, 18)
                .padding(.bottom, 12)
            }
            .scrollDismissesKeyboard(.interactively)
        }
        .background(NT.Colors.surface.ignoresSafeArea())
        .presentationDetents([.height(660), .large])
        .presentationDragIndicator(.hidden)
        .presentationBackground(NT.Colors.surface)
        .presentationCornerRadius(24)
    }

    // MARK: Sections

    private var titleBlock: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text("cant.title").font(NT.Fonts.title2).foregroundStyle(NT.Colors.ink)
            Text(sessionLine).font(NT.Fonts.subheadline).foregroundStyle(NT.Colors.ink2).lineLimit(1)
        }
    }

    private var whySection: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("cant.why").eyebrow()
            BroFlowLayout(spacing: 8) {
                ForEach(CantReason.allCases) { r in
                    SheetChip(title: r.label, isSelected: reason == r) {
                        reason = reason == r ? nil : r
                    }
                }
            }
            noteField
        }
    }

    private var noteField: some View {
        HStack(spacing: 10) {
            TextField("cant.note", text: $note)
                .font(NT.Fonts.body)
                .foregroundStyle(NT.Colors.ink)
                .focused($noteFocused)
                .submitLabel(.done)
                .onChange(of: note) { _, value in
                    if value.count > maxNoteLength { note = String(value.prefix(maxNoteLength)) }
                }
            Text("\(note.count)/\(maxNoteLength)")
                .font(NT.Fonts.footnote)
                .foregroundStyle(NT.Colors.ink3)
                .tabular()
        }
        .padding(.horizontal, 14)
        .frame(height: NT.Size.control)
        .background(NT.Colors.ground, in: RoundedRectangle(cornerRadius: NT.Radius.field, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: NT.Radius.field, style: .continuous).strokeBorder(NT.Colors.hairline, lineWidth: 1))
    }

    private var makeUpSection: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("cant.makeUp").eyebrow()
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 8) {
                    ForEach(makeUpDays, id: \.self) { day in
                        SheetChip(title: BroDerived.weekdayDay(day), isSelected: makeUp == .day(day)) {
                            makeUp = makeUp == .day(day) ? nil : .day(day)
                        }
                    }
                    SheetChip(title: String(localized: "cant.skipThisOne"), isSelected: makeUp == .skip) {
                        makeUp = makeUp == .skip ? nil : .skip
                    }
                }
                .padding(.horizontal, NT.Spacing.screenH)
            }
            .padding(.horizontal, -NT.Spacing.screenH)
        }
    }

    private var summaryBox: some View {
        HStack(alignment: .top, spacing: 10) {
            Image(systemName: "calendar.badge.minus")
                .font(.system(size: 16, weight: .semibold))
                .foregroundStyle(NT.Colors.bad)
                .frame(width: 18, height: 18)
            Text(summaryText)
                .font(NT.Fonts.footnote)
                .foregroundStyle(NT.Colors.ink2)
                .fixedSize(horizontal: false, vertical: true)
        }
        .padding(EdgeInsets(top: 12, leading: 14, bottom: 12, trailing: 14))
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(NT.Colors.ground, in: RoundedRectangle(cornerRadius: 14, style: .continuous))
    }

    private var buttons: some View {
        VStack(spacing: 12) {
            if let partnerName {
                PrimaryButton(title: "cant.send \(partnerName)", isEnabled: !isSending, action: send)
            } else {
                PrimaryButton(title: "cant.sendSolo", isEnabled: !isSending, action: send)
            }
            Button { dismiss() } label: {
                Text("cant.neverMind")
                    .font(NT.Fonts.subheadline)
                    .foregroundStyle(NT.Colors.ink2)
                    .frame(maxWidth: .infinity)
                    .frame(height: NT.Size.control)
                    .contentShape(Rectangle())
            }
            .buttonStyle(PressScale())
        }
    }

    // MARK: Copy

    /// "Friday 18:00 · Push A · Tomek"
    private var sessionLine: String {
        var parts = ["\(Fmt.relativeDay(sessionDay)) \(Fmt.time(minuteOfDay: minuteOfDay))"]
        if let routineName { parts.append(routineName) }
        if let partnerName { parts.append(partnerName) }
        return parts.joined(separator: " · ")
    }

    private var summaryText: String {
        let dayName = Fmt.relativeDay(sessionDay)
        var text: String
        if let partnerName {
            text = String(format: String(localized: "cant.summary"), dayName, partnerName)
            if let day = makeUp?.date {
                let when = "\(Fmt.relativeDay(day)) \(Fmt.time(minuteOfDay: minuteOfDay))"
                text += " " + String(format: String(localized: "cant.summaryMakeUp"), when)
            }
        } else {
            text = String(format: String(localized: "cant.summarySolo"), dayName)
        }
        return text
    }

    // MARK: Send

    private func send() {
        guard !isSending else { return }
        isSending = true
        noteFocused = false
        let day = Calendar.current.startOfDay(for: sessionDay)
        let reasonRaw = reason?.rawValue
        let trimmedNote = note.trimmingCharacters(in: .whitespacesAndNewlines)
        let noteValue: String? = trimmedNote.isEmpty ? nil : trimmedNote
        let makeUpDay = makeUp?.date

        AttendanceService.markMissed(day: day, reason: reasonRaw, note: noteValue, makeUp: makeUpDay, context: modelContext)
        let text = noteValue ?? reason?.label ?? ""
        modelContext.insert(HeadsUp(fromMe: true, kind: .cantMakeIt, text: text, sessionDay: day))
        try? modelContext.save()

        if isPaired {
            let context = modelContext
            Task { await BroShared.service.cantMakeIt(reason: reasonRaw, note: noteValue, makeUpDay: makeUpDay, sessionDay: day, in: context) }
        }
        onDone()
        dismiss()
    }
}

File: android/app/src/main/java/app/notomorrow/feature/bro/CantMakeItSheet.kt
package app.notomorrow.feature.bro

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import app.notomorrow.designsystem.Eyebrow
import app.notomorrow.designsystem.Grabber
import app.notomorrow.designsystem.NT
import app.notomorrow.designsystem.NtIcon
import app.notomorrow.designsystem.NtIcons
import app.notomorrow.designsystem.NtSheet
import app.notomorrow.designsystem.NtShapes
import app.notomorrow.designsystem.NtText
import app.notomorrow.designsystem.NtFlowLayout
import app.notomorrow.designsystem.PrimaryButton
import app.notomorrow.designsystem.SheetChip
import app.notomorrow.designsystem.TabularText
import app.notomorrow.designsystem.ntDismissKeyboardOnScroll
import app.notomorrow.designsystem.pressScale
import app.notomorrow.designsystem.sfIconSize
import app.notomorrow.di.ntViewModel
import app.notomorrow.util.Fmt
import app.notomorrow.util.Localizer
import app.notomorrow.util.NtStrings
import app.notomorrow.util.S
import app.notomorrow.util.rememberNtStrings

/**
 * "Can't make it today" — the port of `CantMakeItSheet` (`Features/Bro/CantMakeItSheet.swift`),
 * opened from the Dashboard and from the Bro tab.
 *
 * Sending writes my `AttendanceRecord(.cancelled)`, the make-up day as planned and a local
 * `HeadsUp(.cantMakeIt)`, then tells the backend through `BroService` whenever I am signed in (the
 * partner hears about it only when paired).
 */
@Composable
fun CantMakeItSheet(onDismiss: () -> Unit, onSent: () -> Unit = {}) {
    val model = ntViewModel { container ->
        CantMakeItViewModel(
            scheduleDao = container.db.scheduleDao(),
            attendanceDao = container.db.attendanceDao(),
            routineDao = container.db.routineDao(),
            workoutDao = container.db.workoutDao(),
            pairingDao = container.db.broPairingDao(),
            headsUpDao = container.db.headsUpDao(),
            attendance = container.attendanceService,
            bro = container.broService,
            strings = NtStrings.from(container.app),
        )
    }
    val state by model.uiState.collectAsStateWithLifecycle()
    val strings = rememberNtStrings()
    val focus = LocalFocusManager.current

    // iOS presents a brand-new sheet every time; the view model outlives it.
    LaunchedEffect(Unit) { model.reset() }

    // `CantMakeItSheet.swift:65` `.presentationCornerRadius(24)`.
    NtSheet(onDismiss = onDismiss, minHeight = SHEET_MIN_HEIGHT, cornerRadius = 24.dp) {
        Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.TopCenter) { Grabber() }
        Column(
            modifier = Modifier
                .fillMaxWidth()
                // SwiftUI's interactive scroll-dismisses-keyboard on the sheet's scroll view.
                .ntDismissKeyboardOnScroll()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = NT.Spacing.screenH)
                .padding(top = 18.dp, bottom = 12.dp),
            verticalArrangement = Arrangement.spacedBy(18.dp),
        ) {
            TitleBlock(state, strings)
            WhySection(
                state = state,
                onToggleReason = model::toggleReason,
                onNote = model::setNote,
                onDone = { focus.clearFocus() },
            )
            MakeUpSection(state, model::toggleMakeUp)
            SummaryBox(state, strings)
            Buttons(
                state = state,
                onSend = {
                    focus.clearFocus()
                    // Swift: `onDone(); dismiss()` — only `send()` runs the closure, never a
                    // "Never mind" or a swipe-down.
                    model.send {
                        onSent()
                        onDismiss()
                    }
                },
                onNeverMind = onDismiss,
            )
        }
    }
}

/** iOS `.presentationDetents([.height(660), .large])`. */
private val SHEET_MIN_HEIGHT = 660.dp

// ─────────────────────────────────────────────────────────────────────────────
// Sections
// ─────────────────────────────────────────────────────────────────────────────

@Composable
private fun TitleBlock(state: CantMakeItUiState, strings: Localizer) {
    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
        NtText(
            text = stringResource(S.cant_title),
            style = NT.Fonts.title2,
            color = NT.Colors.ink,
        )
        NtText(
            text = sessionLine(state, strings),
            style = NT.Fonts.subheadline,
            color = NT.Colors.ink2,
            maxLines = 1,
        )
    }
}

@Composable
private fun WhySection(
    state: CantMakeItUiState,
    onToggleReason: (CantReason) -> Unit,
    onNote: (String) -> Unit,
    onDone: () -> Unit,
) {
    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
        Eyebrow(stringResource(S.cant_why))
        NtFlowLayout(modifier = Modifier.fillMaxWidth(), spacing = 8.dp) {
            CantReason.entries.forEach { reason ->
                SheetChip(
                    title = stringResource(reason.labelRes),
                    selected = state.reason == reason,
                    onClick = { onToggleReason(reason) },
                )
            }
        }
        NoteField(note = state.note, onNote = onNote, onDone = onDone)
    }
}

@Composable
private fun NoteField(note: String, onNote: (String) -> Unit, onDone: () -> Unit) {
    val shape = NtShapes.rounded(NT.Radius.field)
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(NT.Size.control)
            .background(NT.Colors.ground, shape)
            .border(1.dp, NT.Colors.hairline, shape)
            .padding(horizontal = 14.dp),
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(Modifier.weight(1f), contentAlignment = Alignment.CenterStart) {
            BasicTextField(
                value = note,
                onValueChange = onNote,
                modifier = Modifier.fillMaxWidth(),
                textStyle = NT.Fonts.body.copy(color = NT.Colors.ink),
                singleLine = true,
                cursorBrush = SolidColor(NT.Colors.ink),
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Done),
                keyboardActions = KeyboardActions(onDone = { onDone() }),
            )
            if (note.isEmpty()) {
                NtText(
                    text = stringResource(S.cant_note),
                    style = NT.Fonts.body,
                    color = PLACEHOLDER_INK,
                    maxLines = 1,
                )
            }
        }
        TabularText(
            text = "${note.length}/${CantMakeItViewModel.MAX_NOTE_LENGTH}",
            style = NT.Fonts.footnote,
            color = NT.Colors.ink3,
        )
    }
}

/**
 * SwiftUI's `TextField("cant.note", …)` draws its prompt in the system
 * `placeholderText` colour, which is dimmer than `NT.Colors.ink3` (the colour the
 * `0/80` counter beside it uses). Matching iOS keeps the prompt visibly quieter
 * than the counter instead of the two reading as the same grey.
 */
private val PLACEHOLDER_INK = Color(0xFFEBEBF5).copy(alpha = 0.30f)

@Composable
private fun MakeUpSection(state: CantMakeItUiState, onToggle: (MakeUpChoice) -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
        Eyebrow(stringResource(S.cant_makeUp))
        Row(
            modifier = Modifier
                .bleedHorizontally(NT.Spacing.screenH)
                .horizontalScroll(rememberScrollState())
                .padding(horizontal = NT.Spacing.screenH),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            state.makeUpDays.forEach { day ->
                val choice = MakeUpChoice.Day(day)
                SheetChip(
                    title = BroDerived.weekdayDay(day),
                    selected = state.makeUp == choice,
                    onClick = { onToggle(choice) },
                )
            }
            SheetChip(
                title = stringResource(S.cant_skipThisOne),
                selected = state.makeUp == MakeUpChoice.Skip,
                onClick = { onToggle(MakeUpChoice.Skip) },
            )
        }
    }
}

@Composable
private fun SummaryBox(state: CantMakeItUiState, strings: Localizer) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(NT.Colors.ground, NtShapes.rounded(14.dp))
            .padding(horizontal = 14.dp, vertical = 12.dp),
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        verticalAlignment = Alignment.Top,
    ) {
        Box(Modifier.size(18.dp), contentAlignment = Alignment.Center) {
            NtIcon(NtIcons.CalendarBadgeMinus, size = sfIconSize(16f), tint = NT.Colors.bad)
        }
        NtText(
            text = summaryText(state, strings),
            style = NT.Fonts.footnote,
            color = NT.Colors.ink2,
        )
    }
}

@Composable
private fun Buttons(
    state: CantMakeItUiState,
    onSend: () -> Unit,
    onNeverMind: () -> Unit,
) {
    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        val partner = state.partnerName
        PrimaryButton(
            title = if (partner != null) {
                stringResource(S.cant_send_s, partner)
            } else {
                stringResource(S.cant_sendSolo)
            },
            enabled = !state.isSending,
            onClick = onSend,
        )
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(NT.Size.control)
                .pressScale(onClick = onNeverMind),
            contentAlignment = Alignment.Center,
        ) {
            NtText(
                text = stringResource(S.cant_neverMind),
                style = NT.Fonts.subheadline,
                color = NT.Colors.ink2,
                maxLines = 1,
            )
        }
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Copy
// ─────────────────────────────────────────────────────────────────────────────

/** "Friday 18:00 · Push A · Tomek" */
private fun sessionLine(state: CantMakeItUiState, strings: Localizer): String {
    val parts = mutableListOf(
        Fmt.relativeDay(state.sessionDay, strings) + " " + Fmt.time(state.minuteOfDay),
    )
    state.routineName?.let { parts.add(it) }
    state.partnerName?.let { parts.add(it) }
    return parts.joinToString(" · ")
}

private fun summaryText(state: CantMakeItUiState, strings: Localizer): String {
    val dayName = Fmt.relativeDay(state.sessionDay, strings)
    val partner = state.partnerName
        ?: return strings.string(S.cant_summarySolo, dayName)
    var text = strings.string(S.cant_summary, dayName, partner)
    val makeUpDay = state.makeUp?.date
    if (makeUpDay != null) {
        val whenText = Fmt.relativeDay(makeUpDay, strings) + " " + Fmt.time(state.minuteOfDay)
        text += " " + strings.string(S.cant_summaryMakeUp, whenText)
    }
    return text
}
